src/main/java/edu/ucsb/cs156/gauchoride/controllers/DriverChatController.java

```diff
@@ -0,0 +1,151 @@
+import org.springframework.web.bind.annotation.DeleteMapping;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.domain.PageRequest;
+
+import javax.validation.Valid;
+
+import edu.ucsb.cs156.gauchoride.entities.DriverChat;
+import edu.ucsb.cs156.gauchoride.entities.User;
+import edu.ucsb.cs156.gauchoride.repositories.DriverChatRepository;
+import edu.ucsb.cs156.gauchoride.errors.EntityNotFoundException;
+import io.swagger.annotations.Api;
+import io.swagger.annotations.ApiOperation;
+import io.swagger.annotations.ApiParam;
+
+@Api(description= "Driver chat information")
+@RequestMapping("/api/driverchats")
+@RestController
+public class DriverChatController extends ApiController {
+
+    @Autowired
+    DriverChatRepository driverChatRepository;
+
+    @Autowired
+    ObjectMapper mapper;
+
+    @ApiOperation(value = "Get a list of all chat message")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN','ROLE_DRIVER')")
+    @GetMapping("/all")
+    public Iterable<DriverChat> getAllMessages() {
+        return driverChatRepository.findAll();
+    }
+
+
+    @ApiOperation(value = "Get the latest N messages based on query parameters")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN','ROLE_DRIVER')")
+    @GetMapping("/list")
+    public Iterable<DriverChat> getRecentMessage(
+        @ApiParam("limit") @RequestParam int limit
+    ) {
+        Pageable pageable = PageRequest.of(0, limit);
+        return driverChatRepository.findAllByOrderByTimeStampDesc(pageable);
+    }
+
+
+    @ApiOperation(value = "Get a single chat message")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN','ROLE_DRIVER')")
+    @GetMapping("")
+    public DriverChat getMessageById(@ApiParam("id") @RequestParam Long id) {
+        DriverChat message =  driverChatRepository.findById(id)
+        .orElseThrow(()->new EntityNotFoundException(DriverChat.class, id));
+
+        return message;
+    }
+
+
+    @ApiOperation(value="Create a new chat message")
+    @PreAuthorize("hasAnyRole('ROLE_ADMIN','ROLE_DRIVER')")
+    @PostMapping("/post")
+    public DriverChat postNewMessage(
+        @RequestBody @Valid DriverChat driverChat
+    )  throws JsonProcessingException {
+
+        DriverChat savedMessage = driverChatRepository.save(driverChat);
+        return savedMessage;
+    }
+
+
+    @ApiOperation(value="Update a driver's message")
+    @PreAuthorize("hasRole('ROLE_DRIVER')")
```

Comment:
Am I correct in asserting that this cannot currently be tested in practice, since even after merging in @jtmnguyen 's change in PR #8 to enable toggling of the database field in user to make a user a driver, that code is not currently connected to the role system?

If so, then we need a separate PR for that before this is mergeable, becuase if it's not testable, it's not mergeable.

Reply:
That is correct. Currently, there is no role for the driver even if you set the driver field to be true in the database, and I will create another issue for this and merge that issue first.